pkg/runtime/interfaces.go

```diff
@@ -23,64 +23,74 @@ import (
-type ObjectCodec interface {
+// Serializer is the core interface for transforming objects into a serialized format and back.
+// Implementations may choose to perform conversion of the object, but no assumptions should be made.
+type Serializer interface {
+	Encoder
 	Decoder
+}
 
-	// EncodeToVersion convert and serializes an object in the internal format
-	// to a specified output version. An error is returned if the object
-	// cannot be converted for any reason.
-	EncodeToVersion(obj Object, outVersion string) ([]byte, error)
-	EncodeToVersionStream(obj Object, outVersion string, stream io.Writer) error
+// Codec is a Serializer that deals with the details of versioning objects. It offers the same
+// interface as Serializer, so this is a marker to consumers that care about the version of the objects
+// they receive.
+type Codec Serializer
+
+// ParameterCodec defines methods for serializing and deserializing API objects to url.Values and
+// performing any necessary conversion. Unlike the normal Codec, query parameters are not self describing
+// and the desired version must be specified.
+type ParameterCodec interface {
+	// DecodeParameters takes the given url.Values in the specified group version and decodes them
+	// into the provided object, or returns an error.
+	DecodeParameters(parameters url.Values, from unversioned.GroupVersion, into Object) error
+	// EncodeParameters encodes the provided object as query parameters or returns an error.
+	EncodeParameters(obj Object, to unversioned.GroupVersion) (url.Values, error)
 }
 
-// ObjectDecoder is a convenience interface for identifying serialized versions of objects
-// and transforming them into Objects. It intentionally overlaps with ObjectTyper and
-// Decoder for use in decode only paths.
-// TODO: Consider removing this interface?
-type ObjectDecoder interface {
-	Decoder
-	// DataVersionAndKind returns the group,version,kind of the provided data, or an error
-	// if another problem is detected. In many cases this method can be as expensive to
-	// invoke as the Decode method.
-	DataKind([]byte) (unversioned.GroupVersionKind, error)
-	// Recognizes returns true if the scheme is able to handle the provided group,version,kind
-	// of an object.
-	Recognizes(unversioned.GroupVersionKind) bool
+// NegotiatedSerializer is an interface used for obtaining encoders, decoders, and serializers
+// for multiple supported media types.
+type NegotiatedSerializer interface {
+	SupportedMediaTypes() []string
+	SerializerForMediaType(mediaType string, options map[string]string) (Serializer, bool)
+	EncoderForVersion(serializer Serializer, gv unversioned.GroupVersion) Encoder
+	DecoderToVersion(serializer Serializer, gv unversioned.GroupVersion) Decoder
 }
 
 ///////////////////////////////////////////////////////////////////////////////
 // Non-codec interfaces
 
+type ObjectVersioner interface {
+	ConvertToVersion(in Object, outVersion string) (out Object, err error)
```

Comment:
I will change that in the PR that moves type registry out of conversion and
into runtime.  It's coupled enough I believe it should be done at the same
time.

On Thu, Dec 10, 2015 at 9:01 AM, David Eads <notifications@github.com>
wrote:

> In pkg/runtime/interfaces.go
> <https://github.com/kubernetes/kubernetes/pull/17922#discussion_r47228534>
> :
>
> >  }
> >
> >  ///////////////////////////////////////////////////////////////////////////////
> >  // Non-codec interfaces
> >
> > +type ObjectVersioner interface {
> > +	ConvertToVersion(in Object, outVersion string) (out Object, err error)
>
> This should take a GroupVersion, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17922/files#r47228534>.
>